lib/matplotlib/backend_tools.py

```diff
@@ -440,79 +441,30 @@ class ToolEnableNavigation(_ToolEnableNavigation):
     pass
 
 
-class _ToolGridBase(ToolBase):
-    """Common functionality between ToolGrid and ToolMinorGrid."""
-
-    _cycle = [(False, False), (True, False), (True, True), (False, True)]
-
-    def trigger(self, sender, event, data=None):
-        ax = event.inaxes
-        if ax is None:
-            return
-        try:
-            x_state, x_which, y_state, y_which = self._get_next_grid_states(ax)
-        except ValueError:
-            pass
-        else:
-            ax.grid(x_state, which=x_which, axis="x")
-            ax.grid(y_state, which=y_which, axis="y")
-            ax.figure.canvas.draw_idle()
-
-    @staticmethod
-    def _get_uniform_grid_state(ticks):
-        """
-        Check whether all grid lines are in the same visibility state.
-
-        Returns True/False if all grid lines are on or off, None if they are
-        not all in the same state.
-        """
-        if all(tick.gridline.get_visible() for tick in ticks):
-            return True
-        elif not any(tick.gridline.get_visible() for tick in ticks):
-            return False
-        else:
-            return None
-
-
-class ToolGrid(_ToolGridBase):
+class ToolGrid(ToolBase):
     """Tool to toggle the major grids of the figure."""
 
     description = 'Toggle major grids'
     default_keymap = mpl.rcParams['keymap.grid']
 
-    def _get_next_grid_states(self, ax):
-        if None in map(self._get_uniform_grid_state,
-                       [ax.xaxis.minorTicks, ax.yaxis.minorTicks]):
-            # Bail out if minor grids are not in a uniform state.
-            raise ValueError
-        x_state, y_state = map(self._get_uniform_grid_state,
-                               [ax.xaxis.majorTicks, ax.yaxis.majorTicks])
-        cycle = self._cycle
-        # Bail out (via ValueError) if major grids are not in a uniform state.
-        x_state, y_state = (
-            cycle[(cycle.index((x_state, y_state)) + 1) % len(cycle)])
-        return (x_state, "major" if x_state else "both",
-                y_state, "major" if y_state else "both")
-
-
-class ToolMinorGrid(_ToolGridBase):
+    def trigger(self, sender, event, data=None):
+        sentinel = str(uuid.uuid4())  # Prevent triggering the wrong handler.
```

Comment:
I'd add a comment like
~~~
# Trigger classic grid switching by temporarily setting rcParams['keymap.grid']
# to a unique key and sending an appropriate event.
~~~
to explain what we're doing here on the higher level.

Also, is it safe to modify the event here? Now knowing about the implementation, I could imagine e.g. that events are divided in separate subclasses like `MouseEvent`, `KeyPressEvent` and that `key_press_handler()` will ignore `MouseEvent`s. But for some reason this is triggered via a `MouseEvent`.

I may be overly cautious here. From the little I know about the event system, I have the impression that it's rather loosely defined. I'm afraid that this could break if we do refinements on the event system.

Speaking of which, is it possible to test this?

Reply:
Added your clarifying comment.
The event is completely restored immediately after the event is processed, so we should be fine.
There's no tests for now, but helping setting up testing is actually a good reason for this dedupe -- that means less stuff to test twice (between the classic toolbar and the toolmanager).

Reply:
Well, actually I wanted to test this injection logic, which means you'd need to write a test around toolmanager. Independently, one may want a test for the classic trigger as well. 👼 

Anyway, I'm half-way approving without a test, but want to think about it for a bit.